### PR TITLE
[PARO-741] Add a parameter for the user to optionally specify the CivisML version 

### PR DESCRIPTION
**Pull request by sahildshah1**
This PR adds a parameter for the user to optionally specify the CivisML version 

The changes in this PR were modeled after the changes made to the Python API client to use aliases instead of template IDS. See civisanalytics/civis-python#341

### CODE REVIEW

**Comment by sahildshah1** [comment]
@patr1ckm @jacksonllee I ran civis_ml (and predict) interactively

```
library(devtools)
load_all()
```

***civisml_version="foo"***

`gb_model <- civis_ml_gradient_boosting_classifier(iris, "Species",civisml_version="foo")`
produced 
```
Error in get_train_template_id(civisml_version = civisml_version) :
  foo is an invalid CivisML version. Either this version does not exist or you do not have access to this version. Versions accessible to you are: prod, dev, v2.2, v0.5, v1.0, v1.1, v2.0, v2.1 as well as `prod` for the latest production version.
```

***civisml_version="prod"***

`gb_model <- civis_ml_gradient_boosting_classifier(iris, "Species",civisml_version="prod")`
produced 
https://platform.civisanalytics.com/spa/#/models/60983362
and
`predict(gb_model, newdata = iris)`
produced
https://platform.civisanalytics.com/spa/#/scripts/custom/60983511

***civisml_version="dev"***

`gb_model <- civis_ml_gradient_boosting_classifier(iris, "Species",civisml_version="dev")`
produced
https://platform.civisanalytics.com/spa/#/scripts/custom/60983748
and
`predict(gb_model, newdata = iris)`
produced
https://platform.civisanalytics.com/spa/#/scripts/custom/60983993


I didn't test every version of CivisML  https://civis.zendesk.com/hc/en-us/articles/360000260011-CivisML  or every model type https://civisanalytics.github.io/civis-r/articles/civis_ml.html 

Is this sufficient testing of the functions?

**Comment by sahildshah1** [comment]
As noted above, I didn't test every version of CivisML https://civis.zendesk.com/hc/en-us/articles/360000260011-CivisML or every model type https://civisanalytics.github.io/civis-r/articles/civis_ml.html but I did test

***civisml_version="v1.0"***

The training successfully ran:
`gb_model <- civis_ml_gradient_boosting_classifier(iris, "Species",civisml_version="v1.0")`
https://platform.civisanalytics.com/spa/#/scripts/custom/60984495 

But the scoring did not
`predict(gb_model, newdata = iris)` 

```
Error: scripts_get_custom_runs(id = 60984987, run_id = 208384422):
2020-02-06 01:06:36 Queued
2020-02-06 01:06:37 Actual CPU allocated: 256m
2020-02-06 01:06:38 Dedicating resources
2020-02-06 01:06:39 Job is now running
2020-02-06 01:07:52 Failed: The job container failed. Exit code 1
```

The job log reports
```
2020-02-06 01:07:42,130    DEBUG [     8   connectionpool._new_conn           : 818] Starting new HTTPS connection (1): api.civisanalytics.com
2020-02-06 01:07:42,650    DEBUG [     8   connectionpool._make_request       : 395] https://api.civisanalytics.com:443 "GET /endpoints HTTP/1.1" 200 None
2020-02-06 01:07:49,439    DEBUG [     8   connectionpool._new_conn           : 818] Starting new HTTPS connection (1): api.civisanalytics.com
2020-02-06 01:07:49,707    DEBUG [     8   connectionpool._make_request       : 395] https://api.civisanalytics.com:443 "GET /users/me HTTP/1.1" 200 713
2020-02-06 01:07:49,710    DEBUG [     8             main._log_final_exception: 181] Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/civismodel-1.0.8-py3.6.egg/civismodel/main.py", line 108, in cli
    metadata = predict(client=client)
  File "/opt/conda/lib/python3.6/site-packages/civismodel-1.0.8-py3.6.egg/civismodel/main.py", line 275, in predict
    client=client)
  File "/opt/conda/lib/python3.6/site-packages/civismodel-1.0.8-py3.6.egg/civismodel/predict.py", line 233, in run_prediction
    n_jobs = jobs.scale_n_jobs(n_jobs, client=client)
  File "/opt/conda/lib/python3.6/site-packages/civismodel-1.0.8-py3.6.egg/civismodel/configuration/jobs.py", line 24, in scale_n_jobs
    cluster_id = _get_worker_cluster_id(client)
  File "/opt/conda/lib/python3.6/site-packages/civismodel-1.0.8-py3.6.egg/civismodel/configuration/jobs.py", line 58, in _get_worker_cluster_id
    clusters = client.clusters.list_workers()
AttributeError: 'Clusters' object has no attribute 'list_workers'

2020-02-06 01:07:49,710 CRITICAL [     8             main._log_final_exception: 182] 'Clusters' object has no attribute 'list_workers'
2020-02-06 01:07:50,048    DEBUG [     8   connectionpool._make_request       : 395] https://api.civisanalytics.com:443 "POST /files HTTP/1.1" 201 695
2020-02-06 01:07:50,057    DEBUG [     8   connectionpool._new_conn           : 818] Starting new HTTPS connection (1): civis-console.s3.amazonaws.com
2020-02-06 01:07:50,194    DEBUG [     8   connectionpool._make_request       : 395] https://civis-console.s3.amazonaws.com:443 "POST / HTTP/1.1" 204 0
2020-02-06 01:07:50,373    DEBUG [     8   connectionpool._make_request       : 395] https://api.civisanalytics.com:443 "POST /scripts/containers/60984987/runs/208384422/outputs HTTP/1.1" 202 132
2020-02-06 01:07:50,374    DEBUG [     8             main.cli                 : 141] Stored metadata as "model_info.json".
2020-02-06 01:07:50,375    DEBUG [     8             main.cli                 : 157] Writing logs to files endpoint
```

I will dig into the logs to see why the job failed and if it is b/c of this PR

File: R/civis_ml.R [hunk 1]
```diff
@@ -71,6 +71,8 @@
 #'   data for validation, and \code{skip}, which skips the validation step.
 #' @param verbose Optional, If \code{TRUE}, supply debug outputs in Platform
 #'   logs and make prediction child jobs visible.
+#' @param civisml_version Optional, a one-length character vector of the
+#'   CivisML version.  If not provided, the latest version in production is used.
```
Comment (patr1ckm) [inline R/civis_ml.R]: Maybe `default "prod", the latest version in production`

File: R/civis_ml_utils.R [hunk 2]
```diff
@@ -19,13 +19,27 @@ CIVIS_ML_CLASSIFIERS <- c("sparse_logistic",
...
+  id <- civis_ml_template_ids[civis_ml_template_ids$version == civisml_version &
                               civis_ml_template_ids$name == "training", "id"]
 
+  # invalid CivisML version doesn't raise an error
+  if (length(id)==0) {
+
+    stop(paste(civisml_version, "is an invalid CivisML version.",
+      "Either this version does not exist or you do not have access to this version.",
+      "Versions accessible to you are:", paste(unique(civis_ml_template_ids$version),
+        collapse=", "), "as well as `prod` for the latest production version."))
```
[the remark this answers is not in the document]
Comment (patr1ckm) [reply R/civis_ml_utils.R]: +1

Super helpful error message! Thank you!

File: vignettes/civis_ml.Rmd [hunk 3]
```diff
@@ -338,7 +338,15 @@ Many estimators take a `sample_weight` argument. This can be be specified with t
...
-Modeling data must be complete. Any missing values will be imputed with the mean of non-null values in a column. 
+Modeling data must be complete. Any missing values will be imputed with the mean of non-null values in a column.
+
+### CivisML Versions
+
+By default, CivisML uses its latest version in production. Under special
+circumstances, if you would like a specific version (e.g., an older version),
+both `civis_ml` and the `civis_ml_*` functions have the optional parameter
+``civisml_version`` that accepts a string, e.g., ``'v2.3'``
+for CivisML v2.3. Please see [here](https://civis.zendesk.com/hc/en-us/articles/360000260011-CivisML) for the list of CivisML versions.
```
[the remark this answers is not in the document]
Comment (patr1ckm) [reply vignettes/civis_ml.Rmd]: Agreed!

File: R/civis_ml_utils.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (sahildshah1) [reply R/civis_ml_utils.R]: Removed. Thank you for the comment!

**Comment by sahildshah1** [comment]
> Re: the test job failure when you tried `"v1.0"`, I'm 99% sure it has nothing to do with this PR. The error message is `AttributeError: 'Clusters' object has no attribute 'list_workers'`, which is almost certainly because CivisML v1.0 is too old by now. From internal documentation, it looks like the Civis Platform API no longer has any methods for /clusters/workers anymore, which would explain the observed error. If you do want to test this PR, how about `"v2.2"` (the current version in production) instead?

@jacksonllee I tested this PR setting `civisml_version="prod"`  and `civisml_version="dev"` (See the comment above). Is that sufficient?

**Comment by jacksonllee** [comment]
> I tested this PR setting civisml_version="prod" and civisml_version="dev" (Please see #224 (comment).) Is that sufficient?

Yes, this is great! Thanks for running the tests.

LGTM from me -- deferring to @patr1ckm for the official stamp.

**Pull request merged**